src/agents/graph.py

```diff
@@ -0,0 +1,342 @@
+import json
+from collections.abc import Hashable
+from typing import Any, AsyncIterator, Dict, Literal, Protocol  # noqa UP
+
+from langchain_core.messages import AIMessage, HumanMessage
+from langchain_core.output_parsers import PydanticOutputParser
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langgraph.checkpoint.base import BaseCheckpointSaver
+from langgraph.constants import END, START
+from langgraph.graph import MessagesState, StateGraph
+from langgraph.graph.graph import CompiledGraph
+
+from agents.common.agent import Agent
+from agents.common.data import Message
+from agents.common.state import AgentState, Plan, SubTask, UserInput
+from agents.common.utils import EXIT, exit_node, should_exit
+from agents.k8s.agent import K8S_AGENT, KubernetesAgent
+from agents.kyma.agent import KYMA_AGENT, KymaAgent
+from agents.supervisor.agent import COMMON, FINALIZER, SUPERVISOR, SupervisorAgent
+from utils.langfuse import handler
+from utils.logging import get_logger
+from utils.models import Model
+
+logger = get_logger(__name__)
+
+PLANNER = "Planner"
+
+
+def should_continue(state: MessagesState) -> Literal["action", "__end__"]:
+    """Return the next node to execute."""
+    last_message = state["messages"][-1]
+    return "action" if last_message.tool_calls else "__end__"
+
+
+class CustomJSONEncoder(json.JSONEncoder):
+    """Custom JSON encoder."""
```

Comment:
Can you write more comments how does this transfer the obj from which structure to which structure?

Reply:
The expected input and output structure.